Pgp add key storage recommendations

cpaelzer opened the pull request:
### Description

    For a long time everyone has done a lot on their own to keep PGP
    keys safe, and to be fair many already do what I now suggest to
    officially recommend.
    But to step up the integrity of the project we need to start somewhere
    and here that is by outlining clear recommendations that can
    become deeply tested, documented and eventually mandatory.

### Related issue

Well the continuous uncertainty of how to set up keys right and what to do to keep them safe - not an issue tracked here but an issue for sure :-/

### Checklist

- [y] I have read and followed the [Ubuntu Project contributing guide](https://canonical-ubuntu-project.readthedocs-hosted.com/contributors/contribute-docs/)
- [n/a] My pull request is linked to an existing issue (if applicable)
- [y] I have tested my changes, and they work as expected (that is why it does not yet come with a how to use it, that needs me and others to exercise and write it up)

### Additional notes (optional)

I expect that there will be various personal preferences or "but hey you can't enforce" concerns, but you know what.
That is why this is a recommendation for now which still already gets us ahead of the bad former state and hence I'd ask to not drown in bikeshedding or what-if questions. Let us make this a guide that helps everyone to do better and define a suggested standard setup we know to work well.

@s-makin / @rkratky - I'm happy about your review, but please do not land this immediately. After a first round of feedback and polish by close peers I'd want to reach out further to get more feedback and acknowledgement before we land it. I'd let you know when I feel we are ready.

lucistanescu commented on docs/contributors/setup/pgp-key-storage.md:
```diff
@@ -0,0 +1,141 @@
+generated using the on-chip HSM, you would not have a software copy. Therefore
+it's recommended to keep one key in a safe location, e.g. do not keep the pair
+on the same keychain as they would be affected by the same disaster.
+
```
What do you think about also recommending enforcing user-presence (i.e. touch) for the keys. Also, YubiKeys have two modes `on` and `fixed`, only the latter of which is useful when combined with attestation.

lucistanescu commented on docs/contributors/setup/pgp-key-storage.md:
```diff
@@ -0,0 +1,141 @@
+safeguard to recover, but also has no need to be comfortable or quick to reach
+and hence can make use of various extra levels of extra protection against
+physical and digital attacks.
+
```
My usual recommendation is to only ever generate and unlock the main key on an offline live system (and never use that passphrase on any other system).

cpaelzer replied:
That is quite intense but a good hint. I added it as an example of such "extra protection" right here below this paragraph.

lucistanescu commented on docs/contributors/setup/pgp-key-storage.md:
```diff
@@ -0,0 +1,141 @@
+point to have users revisit the setup and current policies.
+They further act as a "dead man's switch" to prevent permanent use of
+compromised or forgotten subkeys.
+
```
What do you think about also recommending the generation of a revocation certificate that can be used if the main key is lost? Obviously, would need to be stored safely, as anyone can publish it.

cpaelzer replied:
yeah, it needs to be stored as safe as the main key IMHO. But then ... you hopefully need it rarely so it does not need to be super comfortable.

lucistanescu commented on docs/contributors/setup/pgp-key-storage.md:
```diff
@@ -0,0 +1,141 @@
+(pgp-encrypted-offline-storage)=
+## Encrypted Offline Storage
+
+A setup would usually consist of a locked-away, offline, main PGP key
```
Maybe clarify that the main key is generated in software?

cpaelzer commented on docs/contributors/setup/pgp-key-storage.md:
```diff
@@ -0,0 +1,141 @@
+becomes questionable with the increasing availability of computational power
+capable of brute-forcing even complex passphrases.
+
+Practical recommendations all derive from the entopy given by the passphrase:
```
Good catch, thanks @jpm-canonical

cpaelzer commented:
FYI about recent updates
- Rebased
- Fixed a automation-reported spellcheck issue
- Accepted typographic fixes - thanks @lucistanescu and @jpm-canonical 
- Integrated the feedback of @lucistanescu and @edoardocanepa

Thank you all for your reviews!

cpaelzer commented on docs/contributors/setup/pgp-key-storage.md:
```diff
@@ -0,0 +1,171 @@
+  operations. This adds a crucial layer of control over who can manage and use
+  the key.
+
+By generating subkeys (see below) directly on the hardware key you furthermore
+gain the ability to create an
+[attestation certificate](https://developers.yubico.com/PGP/Attestation.html)
+can then be published by you and verified by third-parties.
```
Thanks, I've split the sentence to not get overly long.
The second part now is a hopefully more clear "Such a certificate can be published by you to allow verification of your key by third-parties."

cpaelzer commented on docs/contributors/setup/pgp-key-storage.md:
```diff
@@ -0,0 +1,171 @@
+by setting up subkeys for signature/encryption/authentication, which can have
+individual expiry dates.
+
+It is not important to set an expiration on the offline main PGP key, as it
+could easily be used to extend its own expiration date. But on sub-keys this
+can be very effective.
```
I had that initially but got told that you can time warp a machine, extend the expiry date and are out of it (simplified summary) and therefore this is a placebo not really helping much. It is an extra hurdle and we are not saying you "should not" but we mention that it isn't important to do so.

Not changing to this suggestion but taken all others @bdrung - thanks!

cpaelzer commented:
Recent updates

* Rebased to latest head
* Added further suggestions - thanks @bdrung

cpaelzer commented:
Recent updates

* Rebased to latest head
* Added a paragraph to be a platform for potential sufficient alternatives to be added in addition to the main recommendation

basak commented on docs/contributors/setup/pgp-key-storage.md:
```diff
@@ -0,0 +1,205 @@
+To make an example what such extra protection could be like: When the key is
+generated or in the rare cases it is later needed, for example
+to generate new subkeys, it is suggested to do so on an offline (no network)
+live (boot from trusted unmodified medium) system and never use the associate
```
associated

cpaelzer commented:
Recent updates

* add a section about control and ownership (thanks for the discussion on the mailing list)
* refer to the glossary for signing keys
* list known missing aspects visible to the reader
* acknowledge the lack of requirements for alternatives
* Fix time-time: associate -> associated

3v1n0 commented on docs/contributors/setup/pgp-key-storage.md:
```diff
@@ -0,0 +1,244 @@
+  sub-keys you will not need it for daily use.
+* If you prefer memorability: use 6+ [Diceware words](https://en.wikipedia.org/wiki/Diceware)
+  (or a memorable sentence of 20–30 characters).
+* If you use a password manager/geneator: create a random 16+ character password (include upper/lower/digits/symbols).
```
```suggestion
* If you use a password manager/generator: create a random 16+ character password (include upper/lower/digits/symbols).
```

3v1n0 commented on docs/contributors/setup/pgp-key-storage.md:
```diff
@@ -0,0 +1,244 @@
+(pgp-hardware-security-keys)=
+## Hardware Security Keys
+
+To mitigate the elevated risks associated with a compromised PGP key, the Ubuntu
```
Note, you need a recent GPG build (even though the noble version would be enough): https://salsa.debian.org/debian/gnupg2/-/commit/9bf407ad60ec4e03f1769eb715f23b9d2d280244

3v1n0 replied:
Well with TPM you can still have a PIN to sign, however I agree we don't have to list here all the potential supported ways, but a quick mention is not bad either.

cpaelzer replied:
Thanks @paride , yeah helping most by recommending an opinionated way that we believe works well while avoiding "what if too many other options" discussion is just what I wanted. It would be lost in pro/con discussions which I can't resolve for those other options - but with the help of the many people that already provided input I think we can be confident in what I propose.
I left the door open to discuss alternatives separately if anyone wants to truly push for them later on.

@3v1n0 if you are ok with that compromise would you close this discussion as resolved? I do not want to force-close it unless you agree - but try to get all of them closed.

cpaelzer replied:
Wow - leaving my former post up for transparency so you do not go mad when you see your inbox ?!?
The UI didn't refresh and I didn't see your answer, so I decided to ping about it.
And then when I checked later I see the perfect answer to the question I have not yet have asked :-)

Thanks for your input @3v1n0 - resolving this discussion ...

cpaelzer commented on docs/contributors/setup/pgp-key-storage.md:
```diff
@@ -0,0 +1,244 @@
+on the hardware security keys that are used for regular operations.
+That encrypted offline main PGP key is then the final safeguard to recover,
+but also has no need to be comfortable or quick to reach and hence can make
+use of various extra levels of extra protection against physical and digital
```
lol, you are right - let me be extra³ careful

cpaelzer commented on docs/contributors/setup/pgp-key-storage.md:
```diff
@@ -0,0 +1,244 @@
+compromised or forgotten subkeys.
+
+It is important to understand that this expiration is good for {term}`signing keys <Signing Key>`,
+for keys utilized for encryption, for example in confidential mail exchanges,
+then you would usually want to be able to hold on to the key for however long
+you want/need to be able to decrypt whatever was encrypted by others for you
+using its corresponding public key. The recommendation for those is to generate
```
Even if you have software that blocks, you can time warp and extend or create a subkey with extended time. Therefore security told me in a convincing way that expiration on the main key isn't useful.

With your question up I now struggle how to better explain ...
I mean this is in "it is not important to ..." maybe it is ok to not talk in lengths about that aspect?
@paride what do you think?

cpaelzer replied:
Now I got it, in line 130-136 - that was out of a former discussion with people from our security team, but yes you are right - in that case one could extend dates on the sub-keys which is easier. Let me improve on that ...

I'll fuse the two paragraphs with that in mind ...

cpaelzer commented on docs/contributors/setup/pgp-key-storage.md:
```diff
@@ -0,0 +1,244 @@
+we standardize more on it.
+
+Until then, starting with these articles on
+[Using your YubiKey with OpenPGP](https://support.yubico.com/hc/en-us/articles/360013790259-Using-Your-YubiKey-with-OpenPGP)
+and the [YubiKey Guide](https://drduh.github.io/YubiKey-Guide) are a great way
```
added

cpaelzer commented:
Recent updates

* rebased to latest main branch
* Fix misunderstandable phrases (thanks @paride)
* Acknowledge that "secure location" is undefined (thanks @3v1n0)
* Mention alternative HW key options (thanks @3v1n0)

cpaelzer commented on docs/contributors/setup/pgp-key-storage.md:
```diff
@@ -0,0 +1,248 @@
+hardware security keys that support PGP operations. Devices such as some
+YubiKey models, designed with secure elements, offer a significant enhancement
+to purely file-based PGP key protection. We are not selecting a particular
+manufacturer like YUbikey or model, alternatives like
```
Thanks @basak , fixed in the most recent push
